### PR TITLE
add backup option

### PR DESCRIPTION
**Pull request by gregbunce**
Erik and I successfully ran the project via the CLI using a variety of docopt options.  we also added a backup option that was also successfully tested.  here's a sample cli call: 

`python -m sweeper sweep empties Ten_streets_2_bad --workspace=L:\agrc\projects\sweeper\test_data\sweeper_testing.gdb --save-report=C:\Temp --try-fix --backup-to=C:\Temp\testing.gdb`

### CODE REVIEW

File: src/sweeper/backup.py [hunk 1]
```diff
@@ -4,3 +4,19 @@
 backup.py
 A module that creates a gdb if it doesn't exist and inserts a feature class
 '''
+import arcpy 
+
+def backup_data(source_data, out_path):
+    # if the gdb does not exist, then create it
+    if not arcpy.Exists(out_path):
+        out_dir = out_path.rsplit('\\', 1)[0]
+        gdb_name = out_path.rsplit('\\', 1)[1]
```
Comment (steveoh) [inline src/sweeper/backup.py]: I think all of your rsplitting code has bugs

File: src/sweeper/__main__.py [hunk 2]
```diff
@@ -20,14 +20,18 @@
...
 
 def main():
     '''Main entry point for program. Parse arguments and pass to sweeper modules
     '''
 
     args = docopt(__doc__, version='1.0.0')
 
+    # backup file before quality checks
+    if args['--backup-to']:
+        backup.backup_data(args['--workspace'] + '\\' + args['<table_name>'], args['--backup-to'])
```
Comment (steveoh) [inline src/sweeper/__main__.py]: it is not a good practice to join file paths this way. use `os.path.join` instead.

File: src/sweeper/__main__.py [hunk 3]
```diff
@@ -37,12 +41,12 @@ def main():
         print(args)
     elif args['empties']:
         print(args)
-
+        # initialize the class object
```
Comment (steveoh) [inline src/sweeper/__main__.py]: comments should use the `#:` format so you can differentiate between a comment and commented out code.

File: .vscode/settings.json [hunk 4]
```diff
@@ -6,5 +6,5 @@
     "python.formatting.provider": "yapf",
     "python.linting.pylintEnabled": true,
     "python.linting.enabled": true,
-    "python.pythonPath": "C:\\Users\\zbeck\\AppData\\Local\\ESRI\\conda\\envs\\sweeper\\python.exe"
+    "python.pythonPath": "C:\\Users\\gbunce\\AppData\\Local\\Programs\\ArcGIS\\Pro\\bin\\Python\\envs\\sweeper\\python.exe"
```
Comment (steveoh) [inline .vscode/settings.json]: let's stop fighting over the path 

```suggestion
    "python.pythonPath": "%localappdata%\\Programs\\ArcGIS\\Pro\\bin\\Python\\envs\\sweeper\\python.exe"
```

File: src/sweeper/backup.py [hunk 5]
```diff
@@ -4,3 +4,19 @@
 backup.py
 A module that creates a gdb if it doesn't exist and inserts a feature class
 '''
+import arcpy 
+
+def backup_data(source_data, out_path):
+    # if the gdb does not exist, then create it
```
Comment (steveoh) [inline src/sweeper/backup.py]: please add a doc string that tells me what source_data and out_path are. [forklift](https://github.com/agrc/forklift/blob/master/src/forklift/core.py#L65-L74) has great examples

File: src/sweeper/backup.py [hunk 6]
```diff
@@ -4,3 +4,19 @@
...
+    if not arcpy.Exists(out_path):
+        out_dir = out_path.rsplit('\\', 1)[0]
+        gdb_name = out_path.rsplit('\\', 1)[1]
+        arcpy.CreateFileGDB_management(out_dir, gdb_name)
+    
+    # get source_data feature class name
+    fc_name = source_data.rsplit('\\',1)[1]
+
+    # backup the source feature class
+    arcpy.Copy_management(source_data, out_path + '\\' + fc_name)
```
Comment (steveoh) [inline src/sweeper/backup.py]: What happens if the table already exists?

File: src/sweeper/sweepers/empties.py [hunk 7]
```diff
@@ -24,16 +24,16 @@ def sweep(self):
         arcpy.env.workspace = self.workspace
 
         with arcpy.da.SearchCursor(self.table_name, fields) as search_cursor:
-            print("Looping through rows in '{}' ...".format(self.table_name))
+            #print("Looping through rows in '{}' ...".format(self.table_name))
```
Comment (steveoh) [inline src/sweeper/sweepers/empties.py]: remove the print statements or revert the commented out code please.

File: src/sweeper/backup.py (hunk 6 above)
Comment (gregbunce) [reply src/sweeper/backup.py]: if the table exists, then that db will be used as the location for the backup.  is this a better code? 

`    #: check if the db exists
    if not arcpy.Exists(out_path):
        #: if db does not exist, then create it
        out_dir = out_path.rsplit('\\', 1)[0]
        gdb_name = out_path.rsplit('\\', 1)[1]
        arcpy.CreateFileGDB_management(out_dir, gdb_name)
    else:
        #: if db exists, use it as location to load backup file
        pass`

File: .vscode/settings.json (hunk 4 above)
Comment (gregbunce) [reply .vscode/settings.json]: perfect, I copied and added that to my local repo.  i'll push it up with a commit/push today.

File: src/sweeper/__main__.py (hunk 3 above)
Comment (gregbunce) [reply src/sweeper/__main__.py]: makes sense.  thx for the clarification.

File: src/sweeper/backup.py (hunk 5 above)
Comment (gregbunce) [reply src/sweeper/backup.py]: thanks for the example

File: src/sweeper/sweepers/empties.py (hunk 7 above)
Comment (gregbunce) [reply src/sweeper/sweepers/empties.py]: @eneemann are you okay if I remove the print statements, or do you want to handle that?  or do you want to preserve them while debugging stuff?

Comment (eneemann) [reply src/sweeper/sweepers/empties.py]: > @eneemann are you okay if I remove the print statements, or do you want to handle that? or do you want to preserve them while debugging stuff?

Yeah, that's fine.  You can remove them.

Comment (gregbunce) [reply src/sweeper/sweepers/empties.py]: done.

File: src/sweeper/backup.py (hunk 6 above)
Comment (gregbunce) [reply src/sweeper/backup.py]: oops, originally I misread your question.  yup, good question.  should we always append the datetime after the table name?  thoughts?  @steveoh @eneemann

Comment (eneemann) [reply src/sweeper/backup.py]: I think appending the datetime and/or "backup" into the name would work

File: src/sweeper/backup.py (hunk 1 above)
Comment (gregbunce) [reply src/sweeper/backup.py]: it seems to be running okay.  does the syntax look incorrect?

File: src/sweeper/backup.py (hunk 6 above)
Comment (gregbunce) [reply src/sweeper/backup.py]: i added a datetime to the table name.  see what you think.  it's in my most recent commit.

File: src/sweeper/backup.py (hunk 1 above)
Comment (steveoh) [reply src/sweeper/backup.py]: yup you're right

**Pull request merged**